File: solutions/uri/1200/1200.cpp
#include <cstdlib>
#include <iostream>
#include <string>

bool first_in_char;
bool first_pos_char;

struct node {
    char letter;
    node *left;
    node *right;
};

void
    insert_node(node *tree, node *new_node) {
    if (tree->letter > new_node->letter) {
        if (tree->right == NULL) {
            tree->right = new_node;
        } else {
            insert_node(tree->right, new_node);
        }
    } else {
        if (tree->left == NULL) {
            tree->left = new_node;
        } else {
            insert_node(tree->left, new_node);
        }
    }
}

bool
    exist(node *tree, char letter) {
    bool result = false;

    if (tree->letter == letter)
        return true;
    if (tree->right != NULL)
        result = exist(tree->right, letter);
    if (tree->left != NULL)
        result = result || exist(tree->left, letter);

    return result;
}

void
    prefixa(node *tree) {
    std::cout << tree->letter;

    if (tree->right != NULL) {
        std::cout << " ";
        prefixa(tree->right);
    }

    if (tree->left != NULL) {
        std::cout << " ";
        prefixa(tree->left);
    }
}

void
    infixa(node *tree) {
    if (tree->right != NULL)
        infixa(tree->right);

    if (!first_in_char)
        std::cout << " ";
    std::cout << tree->letter;
    first_in_char = false;

    if (tree->left != NULL)
        infixa(tree->left);
}

void
    posfixa(node *tree) {
    if (tree->right != NULL)
        posfixa(tree->right);
    if (tree->left != NULL)
        posfixa(tree->left);

    if (!first_pos_char)
        std::cout << " ";
    std::cout << tree->letter;
    first_pos_char = false;
}

int
    main() {
    char letter;
    std::string command;

    node *tree = new node;
    tree->letter = '-';

    while (std::cin >> command) {
        if (command == "I") {
            std::cin >> letter;

            if (tree->letter == '-') {
                tree->letter = letter;
                continue;
            }

            node *new_node = new node;
            new_node->letter = letter;

            insert_node(tree, new_node);
        } else if (command == "P") {
            std::cin >> letter;

            if (exist(tree, letter)) {
                std::cout << letter << " existe" << std::endl;
            } else {
                std::cout << letter << " nao existe" << std::endl;
            }
        } else if (command == "PREFIXA") {
            prefixa(tree);
            std::cout << std::endl;
        } else if (command == "INFIXA") {
            first_in_char = true;
            infixa(tree);
            std::cout << std::endl;
        } else {
            first_pos_char = true;
            posfixa(tree);
            std::cout << std::endl;
        }
    }

    return 0;
}

File: solutions/beecrowd/1200/1200.py
import sys

add_space_before_letter = True


class Node:
    def __init__(self, letter=None, left=None, right=None):
        self.letter = letter
        self.left = left
        self.right = right

    def insert(self, new_node):
        if not self.letter:
            self.letter = new_node.letter
            return

        if self.letter > new_node.letter:
            if self.right:
                self.right.insert(new_node)
            else:
                self.right = new_node
        else:
            if self.left:
                self.left.insert(new_node)
            else:
                self.left = new_node

    def has(self, letter):
        if self.letter == letter:
            return True

        result = False

        if self.right and self.letter > letter:
            result |= self.right.has(letter)
        if self.left and self.letter < letter:
            result |= self.left.has(letter)

        return result

    def __print(self):
        global add_space_before_letter
        if add_space_before_letter:
            print(" ", end="")
        print(self.letter, end='')
        add_space_before_letter = True

    def __traverse_print(self, traverse_type):
        if traverse_type == 'PREFIX':
            self.__print()
        if self.right:
            self.right.__traverse_print(traverse_type)
        if traverse_type == 'INFIX':
            self.__print()
        if self.left:
            self.left.__traverse_print(traverse_type)
        if traverse_type == 'POSTFIX':
            self.__print()

    def prefix_print(self):
        global add_space_before_letter
        add_space_before_letter = False
        self.__traverse_print(traverse_type='PREFIX')
        print()

    def infix_print(self):
        global add_space_before_letter
        add_space_before_letter = False
        self.__traverse_print(traverse_type='INFIX')
        print()

    def postfix_print(self):
        global add_space_before_letter
        add_space_before_letter = False
        self.__traverse_print(traverse_type='POSTFIX')
        print()


def main():
    global add_space_before_letter
    tree = Node(None, None, None)
    for line in sys.stdin:
        command, *args = (x.strip() for x in line.split(' '))
        match [command, *args]:
            case ['I', letter]:
                tree.insert(Node(letter, None, None))
            case ['P', letter]:
                if tree.has(letter):
                    print(f'{letter} existe')
                else:
                    print(f'{letter} nao existe')
            case ['PREFIXA']:
                tree.prefix_print()
            case ['INFIXA']:
                tree.infix_print()
            case ['POSFIXA']:
                tree.postfix_print()


if __name__ == '__main__':
    main()
